subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRole.java

```diff
@@ -0,0 +1,192 @@
+                        return consumptionDeprecated;
+                    }
+
+                    @Override
+                    public boolean isResolutionDeprecated() {
+                        return resolutionDeprecated;
+                    }
+
+                    @Override
+                    public boolean isDeclarationAgainstDeprecated() {
+                        return declarationAgainstDeprecated;
+                    }
+
+                    @Override
+                    public String describeUsage() {
+                        if (description != null) {
+                            return description;
+                        } else {
+                            return RoleDescriber.describeRole(this);
+                        }
+                    }
+                });
+
+        if (Arrays.asList(ConfigurationRoles.values()).contains(result)) {
+            DeprecationLogger.deprecateBehaviour("Custom configuration roles are deprecated.")
+                    .withAdvice("Use one of the standard roles defined in ConfigurationRoles instead.")
+                    .willBeRemovedInGradle9()
+                    .withUpgradeGuideSection(8, "custom_configuration_roles")
+                    .nagUser();
+        }
+
+        return result;
+    }
+
+    static ConfigurationRole forUsage(boolean consumable, boolean resolvable, boolean declarableAgainst, boolean consumptionDeprecated, boolean resolutionDeprecated, boolean declarationAgainstDeprecated) {
+        return forUsage(RoleDescriber.DEFAULT_CUSTOM_ROLE_NAME, consumable, resolvable, declarableAgainst, consumptionDeprecated, resolutionDeprecated, declarationAgainstDeprecated);
+    }
+
+    static ConfigurationRole forUsage(String name, boolean consumable, boolean resolvable, boolean declarableAgainst, boolean consumptionDeprecated, boolean resolutionDeprecated, boolean declarationAgainstDeprecated) {
+        return forUsage(name, consumable, resolvable, declarableAgainst, consumptionDeprecated, resolutionDeprecated, declarationAgainstDeprecated, null);
+    }
+
+    /**
+     * Create a custom role for the current usage of a given configuration.
+     *
+     * @param configuration the configuration to examine
+     * @return a role with matching its usage characteristics
+     */
+    static ConfigurationRole forConfiguration(ConfigurationInternal configuration) {
+        return forUsage(configuration.getName(), configuration.isCanBeConsumed(), configuration.isCanBeResolved(), configuration.isCanBeDeclaredAgainst(), configuration.isDeprecatedForConsumption(), configuration.isDeprecatedForResolution(), configuration.isDeprecatedForDeclarationAgainst());
+    }
+
+    /**
+     * This static util class hides methods internal to the {@code default} methods of {@link ConfigurationRole} which
+     * can be used to build a human-readable description of the usage a role allows.
+     */
+    abstract class RoleDescriber {
```

Comment:
Why is this class `public` (the default modifier in `interface`)? This seems like something that should be internal/package-private top-level class, if it is supposed to "hide methods internal to [this class]".

Reply:
It has to be, private/protected inner classes aren't allowed inside an interface.

The idea behind the nested types is to work around the lack of private methods allowed in the interface itself.  Any methods on `RoleDescriber` aren't present as default methods in `ConfigurationRole`.  Since `RD` is abstract with a private constructor, it can't ever be instantiated, and works as a "namespace" for shared static functionality (public and private) that would otherwise be exposed.

Could also make `RD` a top-level class, but then it would have to keep at least the same level of access to be usable by `CR`.  As an inner class, it is somewhat more hidden and the location hints that the functionality is only meant to be used by the interface.